pmd-core/src/main/java/net/sourceforge/pmd/RuleSet.java

```diff
@@ -48,6 +51,23 @@
     private Filter<File> filter;
 
     /**
+     * Creates a new RuleSet with the given checksum.
+     * @param checksum A checksum of the ruleset, should change
+     * only if the ruleset was configured differently
+     */
+    public RuleSet(final long checksum) {
+        this.checksum = checksum;
+    }
+    
+    /**
+     * Creates a new RuleSet with a random checksum.
+     * Notice this means the results for this analysis will not be cached.
+     */
+    public RuleSet() {
+        this(new Random().nextLong());
```

Comment:
I tend more towards just defining a `RuleSet` must always be ChecksumAware and we do not allow a way around it. This means, we do not need to mark RuleSets explicitly as ChecksumAware.
Maybe we can make the constructor of `RuleSet` private and allow to create rulesets only through `RuleSetFactory`, which
* either loads a ruleset from an XML file - then we have the checksum
* or creates a ruleset with a single xpath rule - we could create the checksum based on the xpath expression. That would be the XPath CLI use case.

`RuleSets` can stay the way it is - always calculating the checksum based on the checksums of the contained `RuleSet` instances.

The XPathCLI is btw. a very old client - it's purpose seems to be to execute a XPath expression against a source file and display the matches. This is exactly what the Designer is able to do within its GUI. It is not even exposed in the [start script](https://github.com/pmd/pmd/blob/master/pmd-dist/src/main/scripts/run.sh), so we might ditch this XPathCLI altogether.